File: android/src/AndroidPort.java
// SPDX-License-Identifier: GPL-2.0-or-later
// Copyright The XCSoar Project

package org.xcsoar;

import java.io.Closeable;

/**
 * The Java interface of the C++ AndroidPort class.
 */
interface AndroidPort extends AndroidSensor {
  void setListener(PortListener listener);

  void setInputListener(InputListener listener);

  /**
   * @see Port::Drain()
   */
  boolean drain();

  int getBaudRate();
  boolean setBaudRate(int baud);

  /**
   * Write data to the port.  Execution blocks until at least one
   * bytes is written or an error occurs or until the timeout expires.
   *
   * @param data the data to be written
   * @param length the number of bytes to be written
   * @return the number of bytes that were written or -1 on error/timeout
   */
  int write(byte[] data, int length);
}

File: src/Device/Port/AndroidPort.cpp
// SPDX-License-Identifier: GPL-2.0-or-later
// Copyright The XCSoar Project

#include "AndroidPort.hpp"
#include "Android/PortBridge.hpp"

#include <cassert>
#include <stdexcept>

AndroidPort::AndroidPort(PortListener *_listener, DataHandler &_handler,
                         PortBridge *_bridge)
  :BufferedPort(_listener, _handler), bridge(_bridge)
{
  assert(bridge != nullptr);

  bridge->setListener(Java::GetEnv(), _listener);
  bridge->setInputListener(Java::GetEnv(), this);
}

AndroidPort::~AndroidPort() noexcept
{
  assert(bridge != nullptr);

  delete bridge;
}

PortState
AndroidPort::GetState() const noexcept
{
  assert(bridge != nullptr);

  return (PortState)bridge->getState(Java::GetEnv());
}

bool
AndroidPort::Drain()
{
  assert(bridge != nullptr);

  return bridge->drain(Java::GetEnv());
}

unsigned
AndroidPort::GetBaudrate() const noexcept
{
  assert(bridge != nullptr);

  return bridge->getBaudRate(Java::GetEnv());
}

void
AndroidPort::SetBaudrate(unsigned baud_rate)
{
  assert(bridge != nullptr);

  if (!bridge->setBaudRate(Java::GetEnv(), baud_rate))
    throw std::runtime_error{"Failed to set baud rate"};
}

std::size_t
AndroidPort::Write(std::span<const std::byte> src)
{
  assert(bridge != nullptr);

  return bridge->write(Java::GetEnv(), src);
}
